test/bin/generate_common_versions.py

```diff
@@ -0,0 +1,253 @@
+            return False
+
+
+def dnf_repo_is_enabled(repo_name):
+    """
+    :param repo_name: the name of the repository to check
+
+    ``dnf_repo_is_enabled`` checks if the repository is enabled on this system.
+    """
+    try:
+        result = subprocess.run(
+            ["dnf", "repolist"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.DEVNULL,
+            text=True,
+            check=True
+        )
+        return repo_name in result.stdout
+    except subprocess.CalledProcessError:
+        return False
+
+
+def repo_exists(repo_url):
+    """
+    :param repo_url: the URL of a repository
+
+    ``repo_exists`` checks if a URL points to a valid repository.
+    """
+    url = repo_url + "/repodata/repomd.xml"
+    r = requests.get(url)
+    if r.status_code == 404:
+        return False
+    else:
+        return True
+
+
+def get_sub_repo(minor):
+    """
+    :param minor: the minor version, e.g. 19 for 4.19
+
+    ``get_sub_repo`` returns the name of the subscription repository for the specified
+    minor version if the repository provides the microshift package, otherwise returns None.
+    """
+    try:
+        repo = f"rhocp-4.{minor}-for-rhel-9-{ARCH}-rpms"
+        _ = subprocess.run(
+                ['sudo', 'dnf', 'repoquery', 'microshift', '--quiet',
+                 '--queryformat', '%{version}-%{release}', '--repo', repo],
+                capture_output=True,
+                text=True,
+                check=True
+            )
+
+        return repo
+
+    except subprocess.CalledProcessError:
+        return None
+
+
+def get_release_repo(minor):
+    """
+    :param minor: the minor version, e.g. 19 for 4.19
+
+    ``get_release_repo`` returns the repository for the specified minor version.
+    It looks for the 'rhocp' stream, release candidate and engineering candidate,
+    in that order. If none of these repositories are available, returns empty
+    double quotes (to be used as empty string in bash).
+    """
+    repo = get_sub_repo(minor)
+    if repo is not None and provides_pkg(repo, "microshift"):
+        return f'"{repo}"'
+
+    rc = get_candidate_repo(minor, dev_preview=False)
+    if repo_exists(rc) and provides_pkg(rc, "microshift"):
+        return f'"{rc}"'
+
+    ec = get_candidate_repo(minor, dev_preview=True)
+    if repo_exists(ec) and provides_pkg(ec, "microshift"):
+        return f'"{ec}"'
+
+    return '""'
+
+
+def get_release_version(repo):
+    """
+    :param repo: the name or the URL of the repo, enclosed in double quotes
+
+    ``get_release_version`` returns a string to be used as a bash variable,
+    with a call to the right function depending on whether the `repo` param
+    is a URL or not. if neither, returns empty double quotes.
+    """
+    if repo.startswith("\"rhocp"):
+        return f'"$(get_vrel_from_rhsm {repo})"'
+    elif repo.startswith("\"https"):
+        return f'"$(get_vrel_from_beta {repo})"'
```

Comment:
Ideally, we want them to still reference variables, not urls.

That is:
```
PREVIOUS_RELEASE_REPO="https://mirror.openshift.com/pub/openshift-v4/$(uname -m)/microshift/ocp/latest-4.19/el9/os"
PREVIOUS_RELEASE_VERSION="$(get_vrel_from_beta "${PREVIOUS_RELEASE_REPO}")"
```
instead of
```sh
PREVIOUS_RELEASE_REPO="https://mirror.openshift.com/pub/openshift-v4/aarch64/microshift/ocp/latest-4.19/el9/os"
PREVIOUS_RELEASE_VERSION="$(get_vrel_from_beta "https://mirror.openshift.com/pub/openshift-v4/aarch64/microshift/ocp/latest-4.19/el9/os")"
```